### PR TITLE
[KYUUBI #1978] Support NEGOTIATE/BASIC authorization for restful frontend service

### LINKED ISSUE #1978: [FEATURE] Kyuubi Restful API Kerberos authentication support
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues?q=is%3Aissue) and found no similar issues.

### Describe the feature

Support the kerberos authentication for restful front end server.

### Motivation

_No response_

### Describe the solution

_No response_

### Additional context

_No response_

### Are you willing to submit PR?

- [X] Yes I am willing to submit a PR!

### PR DESCRIPTION
**Pull request by turboFei**
### _Why are the changes needed?_

to close #1978 

Support both NEGOTIATE and BASIC authentication for restful frontend service.

At first, I involve two auth schemes, FYI [hadoop/HttpConstants.java](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project%2Fhadoop-auth%2Fsrc%2Fmain%2Fjava%2Forg%2Fapache%2Fhadoop%2Fsecurity%2Fauthentication%2Fserver%2FHttpConstants.java)
- BASIC: mapping to NOSASL, NONE, LDAP, CUSTOM authentication
- NEGOTIATE: mapping to KERBEROS authentication

BTW, hadoop also supports `Digest` auth scheme.

Two authentication handlers:
- BasicAuthenticationHandler(reuse existing passwdAuthenticationProvider(LDAP/CUSTOM))
- KerberosAuthenticationHandler(refer [hadoop/KerberosAuthenticationHandler.java](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project%2Fhadoop-auth%2Fsrc%2Fmain%2Fjava%2Forg%2Fapache%2Fhadoop%2Fsecurity%2Fauthentication%2Fserver%2FKerberosAuthenticationHandler.java) and [hadoop/KerberosUtil.java](https://github.com/apache/hadoop/blob/trunk/hadoop-common-project%2Fhadoop-auth%2Fsrc%2Fmain%2Fjava%2Forg%2Fapache%2Fhadoop%2Fsecurity%2Fauthentication%2Futil%2FKerberosUtil.java))

#### AuthenticationFilter
- redirect the http request according to the authentication scheme specified in `Authorization` header.

For basic authentication
The value of Authorization header is `BASIC ${encodeBase64(user + ":" + password)}`.
Then using the passwdAuthenticationProvider to authenticate `user` and `password`.

For SPNEGO(kerberos) authentication.
The value of Authorization header is `NEGOTIATE ${encodeBase64(clientToken)}`.
Then checking whether the clientToken is valid.

And currently, the AuthenticationFilter take affect for all the requests with pathSpec `/api/*`.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by turboFei** [comment]
almost ready

Todo: add ut for negotiate with valid kerberos token

File: kyuubi-server/src/test/scala/org/apache/kyuubi/operation/KyuubiRestAuthenticationSuite.scala [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.operation
+
+import javax.servlet.http.HttpServletResponse
+
+import org.apache.commons.codec.binary.Base64
```
Comment (pan3793) [inline kyuubi-server/src/test/scala/org/apache/kyuubi/operation/KyuubiRestAuthenticationSuite.scala:22]: Prefer to use JDK Base64

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiRestFrontendService.scala [hunk 2]
```diff
@@ -49,7 +54,10 @@ class KyuubiRestFrontendService(override val serverable: Serverable)
   }
 
   private def startInternal(): Unit = {
-    server.addHandler(ApiRootResource.getServletHandler(this))
+    val contextHandler = ApiRootResource.getServletHandler(this)
+    val holder = new FilterHolder(new AuthenticationFilter(conf))
+    contextHandler.addFilter(holder, "/*", EnumSet.allOf(classOf[DispatcherType]))
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiRestFrontendService.scala:59]: Do we need to add the authentacation filter to "/*"? Maybe we need to give some exceptions, e.g. "/healthy-check", "/metrics/**"

File: kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/KerberosUtil.scala [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+import java.io.UnsupportedEncodingException
+import java.nio.ByteBuffer
+import java.nio.charset.IllegalCharsetNameException
+
+import org.ietf.jgss.{GSSException, Oid}
+
+object KerberosUtil {
+  val GSS_SPNEGO_MECH_OID: Oid = getNumericOidInstance("1.3.6.1.5.5.2")
+  val GSS_KRB5_MECH_OID: Oid = getNumericOidInstance("1.2.840.113554.1.2.2")
+  val NT_GSS_KRB5_PRINCIPAL_OID: Oid = getNumericOidInstance("1.2.840.113554.1.2.2.1")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/service/authentication/KerberosUtil.scala:29]: Would you please add some comments on them? Looks like magic things.

File: dev/dependencyList [hunk 4]
```diff
@@ -17,6 +17,7 @@
 
 aopalliance-repackaged/2.6.1//aopalliance-repackaged-2.6.1.jar
 classgraph/4.8.95//classgraph-4.8.95.jar
+commons-codec/1.15//commons-codec-1.15.jar
```
Comment (pan3793) [inline dev/dependencyList:20]: For Base64? Please use JDK Base64 instead

**Comment by yaooqinn** [comment]
Can we explain a bit about your PR and the technologies you introduced? It's hard to get start and go detail review.

**Comment by turboFei** [comment]
have updated in the pr description

**Comment by yaooqinn** [comment]
can we move all the new classes to Server modudle ？

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 5]
```diff
@@ -197,6 +197,19 @@ object KyuubiConf {
     .stringConf
     .createOptional
 
+  val SERVER_SPNEGO_KEYTAB: ConfigEntry[Option[String]] = buildConf("spnego.keytab")
+    .doc("Keytab file for SPNego principal")
+    .version("1.5.0")
+    .fallbackConf(SERVER_KEYTAB)
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:203]: why it is not optional？

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 6]
```diff
@@ -197,6 +197,19 @@ object KyuubiConf {
     .stringConf
     .createOptional
 
+  val SERVER_SPNEGO_KEYTAB: ConfigEntry[Option[String]] = buildConf("spnego.keytab")
+    .doc("Keytab file for SPNego principal")
+    .version("1.5.0")
```
Comment (yaooqinn) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:202]: 1.6

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/authentication/AuthSchemes.scala [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.server.authentication
+
+object AuthSchemes extends Enumeration {
+  type AuthScheme = Value
+
+  val BASIC, NEGOTIATE = Value
```
[the remark this answers is not in the document]
Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/authentication/AuthSchemes.scala:23]: It seems they are the same.
```
Like Basic, Digest access authentication verifies that both parties
   to a communication know a shared secret (a password); unlike Basic,
   this verification can be done without sending the password in the
   clear, which is Basic's biggest weakness. As with most other
   authentication protocols, the greatest sources of risks are usually
   found not in the core protocol itself but in policies and procedures
   surrounding its use.
```

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/authentication/AuthSchemes.scala [hunk 8]
```diff
@@ -0,0 +1,24 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.server.authentication
+
+object AuthSchemes extends Enumeration {
```
[the remark this answers is not in the document]
Comment (turboFei) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/authentication/AuthSchemes.scala:20]: prefer: http. authentication.

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
> The diff coverage is 51.62%.

can we target to 80%+

**Comment by turboFei** [comment]
sure, will add UT.

**Comment by turboFei** [comment]
>  almost ready
>  Todo: add ut for negotiate with valid kerberos token

Have added ut for negotiate with valid kerberos token

**Comment by turboFei** [comment]
image

**Comment by turboFei** [comment]
thanks, merging to master

**Pull request closed without merging**